cephfs/file.go

```diff
@@ -0,0 +1,143 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <stdlib.h>
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+import (
+	"errors"
+	"unsafe"
+)
+
+const (
+	// seek constants. Using private constants allows us avoid
+	// "leaking" unnecessary details in the docs.
+	seekSet = int(C.SEEK_SET)
+	seekCur = int(C.SEEK_CUR)
+	seekEnd = int(C.SEEK_END)
+
+	// SeekSet is used with Seek to set the absolute position in the file.
+	SeekSet = seekSet
+	// SeekCur is used with Seek to position the file relative to the current
+	// position.
+	SeekCur = seekCur
+	// SeekEnd is used with Seek to position the file relative to the end.
+	SeekEnd = seekEnd
+)
+
+var (
+	errInvalidSeekWhence = errors.New("invalid whence value for seek")
+)
+
+// File represents an open file descriptor in cephfs.
+type File struct {
+	mount *MountInfo
+	fd    C.int
+}
+
+// Open a file at the given path. The flags are the same os flags as
+// a local open call. Mode is the same mode bits as a local open call.
+//
+// Implements:
+//  int ceph_open(struct ceph_mount_info *cmount, const char *path, int flags, mode_t mode);
+func (mount *MountInfo) Open(path string, flags int, mode uint32) (*File, error) {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+	ret := C.ceph_open(mount.mount, cPath, C.int(flags), C.mode_t(mode))
+	if ret < 0 {
+		return nil, getError(ret)
+	}
+	return &File{mount: mount, fd: ret}, nil
+}
+
+// Close the file.
+//
+// Implements:
+//  int ceph_close(struct ceph_mount_info *cmount, int fd);
+func (f *File) Close() error {
+	return getError(C.ceph_close(f.mount.mount, f.fd))
+}
+
+// read directly wraps the ceph_read call. Because read is such a common
+// operation we deviate from the ceph naming and expose Read and ReadAt
+// wrappers for external callers of the library.
+//
+// Implements:
+//  int ceph_read(struct ceph_mount_info *cmount, int fd, char *buf, int64_t size, int64_t offset);
+func (f *File) read(buf []byte, offset int64) (int, error) {
+	bufptr := (*C.char)(unsafe.Pointer(&buf[0]))
+	ret := C.ceph_read(
+		f.mount.mount, f.fd, bufptr, C.int64_t(len(buf)), C.int64_t(offset))
+	if ret < 0 {
+		return int(ret), getError(ret)
+	}
+	return int(ret), nil
+}
+
+// Read data from file. Up to len(buf) bytes will be read from the file.
+// The number of bytes read will be returned.
+func (f *File) Read(buf []byte) (int, error) {
+	// to-consider: should we mimic Go's behavior of returning an
+	// io.ErrShortWrite error if write length < buf size?
+	return f.read(buf, -1)
+}
+
+// ReadAt will read data from the file starting at the given offset.
+// Up to len(buf) bytes will be read from the file.
+// The number of bytes read will be returned.
+func (f *File) ReadAt(buf []byte, offset int64) (int, error) {
+	return f.read(buf, offset)
```

Comment:
AFAICT, Go's built-in ReadAt is implemented using pread, with some layers of indirection, but I'm fairly confident it is really using pread syscall.
I can look into making sure the EOF behavior matches.

Reply:
Ok, but if it does use `pread()`, then you should not use `cephfs_read()`, I think.

Reply:
I think your concern is that ceph_read with an offset value will move the internal file position?  I don't remember seeing it documented in the header but AFAIK if -1 is passed it will move the position but if the offset is non-negative it does not change the file position. Just to be sure I checked the ceph sources:
You can see this at https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L9157 and https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L9249

If you want I can even add a test to double check that calling ReadAt doesn't change the file position if mixed calls to ReadAt and Read occur (not that I would encourage doing that combo). A test would also show a failure if ceph ever changed this behavior in the future.

If this is not your concern please let me know and clarify why you think this is an issue.

Reply:
My concern is that a 'normal' Go `ReadAt()` may not use the file position, but that this implementation by using `ceph_read()` does. It would be unexpected for developers that want to convert their application to use go-ceph. This kind of difference in behaviour can be the cause of very difficult to track down bugs.

Reply:
I thought my previous comment clarified this but I'll restate: I examined the ceph source code and ceph_read does not use/change the file position when an offset >0 is used. Thus the ReadAt wrapper will not use/change the file position. To assuage any concerns I can add additional validation of the offset to ReadAt to ensure it it >0. I will also add a test to validate that ReadAt does not change the file position when mixing calls to Read and ReadAt.
Will these alleviate your concerns?